**Coursebook: Python for Data Analysts**

This notebook was made based on main materials `1_Python_for_Data_Analysts.ipynb`

Version: Qoppa - September 2021
___


## Top-Down Approach 

The coursebook is part of the **Data Analytics Specialization** offered by [Algoritma](https://algorit.ma). It takes a more accessible approach compared to Algoritma's core educational products, by getting participants to overcome the "how" barrier first, rather than a detailed breakdown of the "why". 

This translates to an overall easier learning curve, one where the reader is prompted to write short snippets of code in frequent intervals, before being offered an explanation on the underlying theoretical frameworks. Instead of mastering the syntactic design of the Python programming language, then moving into data structures, and then the `pandas` library, and then the mathematical details in an imputation algorithm, and its code implementation; we would do the opposite: Implement the imputation, then a succinct explanation of why it works and applicational considerations (what to look out for, what are assumptions it made, when _not_ to use it etc).

For the most part, experience in Python programming is good to have but not required. Familiarity with data manipulation and data structures in a different programming language a welcome addition but again, not required.

## Training Objectives

This coursebook is intended for participants new to the world of data analysis and / or programming. No prior programming knowledge is assumed. 

The coursebook focuses on:
- Introduction to the `pandas` library. 
- Introduction to `DataFrame`  
- Data Types
- Exploratory Data Analysis I
- Indexing and Subsetting

The final part of this course is a Graded Asssignment, where you are expected to apply all that you've learned on a new dataset, and attempt the given questions.

# Environment

**Cara membuat virtual environment baru:**

1. Membuka Anaconda Prompt

2. Buat environment dengan nama `ENV_NAME`:
    ```
    conda create -n <ENV_NAME> python=<PYTHON_VERSION>
    ```
    Contoh: `conda create -n qoppa_da python=3.9`


3. Mengaktifkan (pindah) environment:
    ```
    conda activate <ENV_NAME>
    ```
    Contoh: `conda activate qoppa_da`
    

4. Memasang kernel di dalam environment:
    ```
    pip install ipykernel
    python -m ipykernel install --user --name=<ENV_NAME>
    ```

5. Cara instalasi package:
    ```
    pip install <PACKAGE_NAME>
    ```
    atau
    ```
    conda install <PACKAGE_NAME>
    ```
    Contoh: `pip install pandas`

**Perintah `conda` yang sering digunakan:**

- `conda env list`: cek daftar **environment** pada Anaconda
- `conda list`: cek daftar **package** di dalam sebuah environment

Referensi perintah lain: https://docs.conda.io/projects/conda/en/latest/commands.html

## Dive Deeper

Replikasilah `jupyter_notebook_practice.html` menggunakan notebook kosongan dan submit file `.ipynb` beserta `.html` ke assignment Google Classroom.

## ❓ Knowledge Check: Summary Day 1

Pilihlah jawaban yang tepat dengan memberikan tanda centang pada kotak. Jawaban bisa lebih dari satu.

1. Tools berikut yang merupakan Integrated Development Environment (IDE) adalah ...

    - [X] Jupyter Notebook
    - [ ] Anaconda --> data science toolboc
    - [ ] Python --> bahasa pemrograman
    - [X] Visual Studio Code (VSCode)

2. Berikut yang merupakan Package Manager untuk instalasi package/library adalah ...

    - [ ] kernel
    - [X] conda 
    - [X] pip 
    - [ ] environment

3. Kita perlu menggunakan ...... agar sebuah virtual environment dapat digunakan pada Jupyter Notebook.

    - [X] kernel 
    - [ ] package manager --> untuk install library/package
    - [ ] library / package
    - [ ] IDE

4. Berikut adalah tujuan kita membuat virtual environment, **KECUALI** ...

    - [ ] Kolaborasi (sharing environment) --> betul
    - [ ] Deliver/Deploy aplikasi ke user --> betul
    - [ ] Mengisolasi package beserta versi-nya --> betul
    - [X] Virtual environment harus dibuat agar Python dapat dijalankan --> salah, python bisa dijalankan tanpa harus membuat virtual environment baru

# Introduction to Jupyter Notebook

## Markdown Cell and Code Cell

Kumpulan shortcut: **CTRL + SHIFT + P**

Tipe cell dalam notebook:
1. Markdown 
2. Code

This is markdown cell. You can write a formatted text such as **bold** or *italic*. You can even write mathematical formula such 

\begin{equation}
f(x) = \frac{e^{-x}}{(1+e^{-x})}
\end{equation}

## Command Mode and Edit Mode

Mode cell dalam notebook:
1. Command mode (cell berwarna BIRU)
    - `a` : add cell above 
    - `b` : add cell below 
    - `d` + `d` : delete selected cell 
    - `x` : cut selected cell 
    - `c` : copy selected cell 
    - `v` : paste selected cell 
    - `z` : undo 
    - `m` : change cell type to markdown 
    - `y` : change cell type to code
    - `enter` : enter Edit Mode
    - `h` : show keyboard shortcuts


2. Edit mode (cell berwarna HIJAU)
    - `Ctrl + Enter`: eksekusi satu cell
    - `Esc`: Mengubah edit mode menjadi command mode

# Introduction to Python


## Variables and Keywords

**Variable** adalah sebuah nama yang dipakai untuk menunjukkan sebuah nilai. Tanda `=` dipakai untuk membuat variable baru. Proses ini sering disebut sebagai **assignment**.


In [1547]:
# Variabel
ini_variabel = ['Daniel', 'Syahputra']

Python adalah bahasa pemrograman yang **case-sensitive** sehingga penamaan variable menjadi hal yang perlu diperhatikan. Misal penulisan variable `activity` dengan awalan huruf `a` kecil berbeda dengan yang diawali dengan huruf `A` besar.

In [1548]:
activity = "Something"
Activity = "Something"

Berikut beberapa anjuran dalam memberikan nama variable pada Python:
- Menggunakan kombinasi dari huruf kapital (A-Z), huruf nomina (a-z), angka (0-9).
- Special character `!, $ , &, dll` tidak dapat digunakan dalam penamaan variabel.
- Tidak boleh menggunakan angka di awal.
- Tidak boleh menggunakan keyword pada Python
- Bersifat case-sensitive sehingga penamaan variable `algoritma`, `ALGORITMA`, dan `Algoritma` adalah 3 variable yang berbeda

**Keywords** adalah kata kunci yang sudah ditetapkan oleh Python sebagai nama yang tidak bisa dipakai baik untuk penamaan fungsi, variabel, dan lainnya. Keyword ditulis dalam lower-case (huruf kecil semua) kecuali keyword `True`, `False`, dan `None`. Sejauh ini (Python 3.8) keyword yang ada pada Python adalah sebagai berikut:

In [1549]:
# Cek daftar keyword
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

## Python Data Types

### String

Python mewakili string apa pun sebagai objek `str`. Ada beberapa cara untuk membuat nilai string:

- menggunakan `''` (yaitu: `'cyber punk 2077'`)
- menggunakan `""` (yaitu : `"Hari Jum'at"`)
- menggunakan `'''` atau `"""` (yaitu: `'''Andi berkata "Jum'at Bersih"'''`)

### Number

Untuk menyimpan number, python memiliki dua tipe data asli yang disebut `int` dan `float`.
- `int` digunakan untuk menyimpan bilangan bulat (yaitu: 1,2,-3)
- `float` digunakan untuk menyimpan bilangan real (yaitu: 0.7, -1.8, -1000.0).

**Operasi Angka** \
Operator Aritmatika:
- `+` - Penambahan
- `-` - Pengurangan
- `*` - Perkalian
- `/` - Divisi
- `//` - Pembagian Putaran
- `%` - Modul
- `**` - Eksponen

Operator Perbandingan:
- `<` - Lebih kecil dari (yaitu : a < b)
- `<=` - Lebih kecil atau sama dengan (yaitu : a <= b)
- `>` - Lebih besar dari (yaitu: a > b)
- `>=` - Lebih besar atau sama dengan (yaitu: a >= b)
- `==` - Sama dengan (yaitu: a == b)
- `!=` - Tidak Sama dengan (yaitu: a != b)

### Boolean

Boolean menyimpan nilai yang sangat sederhana dalam komputer dan pemrograman, `True` atau `False`.

**Operasi Boolean** \
Python menyediakan operator logika seperti:
- and (yaitu: a and b)
- or (yaitu: a or c)
- not (yaitu: not a)

### List

`list` digunakan untuk menyimpan beberapa nilai dalam python. Untuk membuatnya, cukup letakkan nilai di dalam tanda kurung (yaitu: `x = [1,2,3]` )

**Operasi List**
- `x.append(a)` : tambahkan a ke x
- `x.remove(a)` : hapus a dari x

Selain operator yang dikenal sebelumnya, salah satu list yang paling berguna adalah dengan menerapkan fungsi agregasi seperti:
- `len(x)` : ekstrak panjang daftar
- `a in b` : memeriksa apakah nilai `a` ada di objek daftar `b`
- `max(x)` : mendapatkan nilai tertinggi dalam x
- `sum(x)` : mendapatkan jumlah nilai dalam x

Operasi lain yang harus diketahui dalam daftar adalah pengindeksan:
- `x[i]` : mengakses elemen ke-i dari x

# Introduction to Library (Package)

Ketika ingin menggunakan suatu library langkah yang perlu dilakukan adalah:

1. Install library yang digunakan
2. Import/panggil library yang dibutuhkan

## Import a Library

In [1550]:
import pandas as pd


Cara untuk melihat versi perpustakaan setelah dimuat adalah dengan memanggilnya *atribut versi*. Namun **library yang berbeda mungkin memiliki gaya penamaan atribut yang berbeda**.

In [1551]:
# check versi pandas
pd.__version__

'1.3.4'

> Notes: attribute `.__version__` dilakukan untuk melihat versi dari `pandas`

## Install / Update a Library

Jika Anda menggunakan python dari distribusi Anaconda, kemungkinan besar Anda memiliki **2 opsi** untuk menginstal pustaka dari :
1. **pip** \
pip adalah manajer paket untuk Python. Anda dapat menggunakan pip untuk menginstal paket dari [Python Package Index](https://pypi.org/) dan indeks lainnya.
2. **conda** \
conda adalah manajer paket yang disediakan oleh Anaconda. Anda dapat menggunakan conda untuk menginstal paket dari [Anaconda Package Repository](https://anaconda.org/anaconda/repo).

# Pandas Library
## Working with DataFrame

`pandas` adalah library yang powerful sebagai tools analisis data dan struktur pada Python. Dengan `pandas`, mengolah data menjadi mudah karena disediakan salah satu objek bernama **DataFrame**. Dengan dataframe kita dapat membaca sebuah file, mengolah suatu data dengan menggunakan operasi seperti join, distinct, group by, agregasi, dan teknik lainnya.

> Lebih lengkapnya silahkan kunjungi [official documentation](https://pandas.pydata.org/)

Untuk menggunakan `pandas`, kita perlu import terlebih dahulu library dengan cara berikut ini:

Semua method pada `pandas` dapat dipanggil dengan syntax seperti: `pandas.function_name()`. Langkah pertama yang akan kita lakukan adalah membaca data. Kita dapat menggunakan method `.read_csv()` untuk membaca sebuah file dengan format `.csv`.

In [1552]:
# code ini untuk mengatur pemisah angka ribuan
pd.options.display.float_format = '{:,.3f}'.format

In [1553]:
# read data
rice = pd.read_csv('../materi_utama/data_input/rice.csv', index_col=0)
rice

receipt_id  receipts_item_id     purchase_time category sub_category  \
1         9622257          32369294   7/22/2018 21:19     Rice         Rice   
2         9446359          31885876   7/15/2018 16:17     Rice         Rice   
3         9470290          31930241   7/15/2018 12:12     Rice         Rice   
4         9643416          32418582    7/24/2018 8:27     Rice         Rice   
5         9692093          32561236   7/26/2018 11:28     Rice         Rice   
...           ...               ...               ...      ...          ...   
11996     5760491          17555486  12/15/2017 21:06     Rice         Rice   
11997     5598782          16999147   12/2/2017 14:12     Rice         Rice   
11998     5735850          17434503  12/13/2017 19:17     Rice         Rice   
11999     5678748          17317935   12/8/2017 22:04     Rice         Rice   
12000     5702411          17341559  12/10/2017 17:04     Rice         Rice   

            format  unit_price  discount  quantity yearmonth  
1      supermarket 128,000.000         0         1   2018-07  
2       minimarket 102,750.000         0         1   2018-07  
3      supermarket  64,000.000         0         3   2018-07  
4       minimarket  65,000.000         0         1   2018-07  
5      supermarket 124,500.000         0         1   2018-07  
...            ...         ...       ...       ...       ...  
11996  supermarket 128,000.000         0         1   2017-12  
11997   minimarket  64,000.000         0         1   2017-12  
11998  supermarket  59,990.000      3000         1   2017-12  
11999   minimarket  59,500.000      3000         1   2017-12  
12000   minimarket  59,500.000      3000         1   2017-12  

[12000 rows x 10 columns]

In [1554]:
rice.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 1 to 12000
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   receipt_id        12000 non-null  int64  
 1   receipts_item_id  12000 non-null  int64  
 2   purchase_time     12000 non-null  object 
 3   category          12000 non-null  object 
 4   sub_category      12000 non-null  object 
 5   format            12000 non-null  object 
 6   unit_price        12000 non-null  float64
 7   discount          12000 non-null  int64  
 8   quantity          12000 non-null  int64  
 9   yearmonth         12000 non-null  object 
dtypes: float64(1), int64(4), object(5)
memory usage: 1.0+ MB


Kita coba set kolom pertama `Unnamed: 0` menjadi index dari dataframe, dengan cara menambahkan **parameter** `index_col=0`.

In [1555]:
# code here


Intuisi dari parameter `index_col` pada `read_csv()` adalah menjadikan kolom pada Dataframe sebagai index pada baris. Berikut beberapa nilai yang dapat ditampung oleh parameter `index_col`.
- (0, 1, 2, dst) : 
- `'nama_kolom'` : 
- `False` : 

**📌 Additional Information:**

1. Python menggunakan sistem **zero based indexing** yang berarti, index pada python dimulai dari angka 0.
2. Terdapat beberapa parameter dalam method `.read_csv()`, dokumentasi selengkapnya ada [di sini](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)

### Dive Deeper

1. Baca kembali data `rice` menggunakan perintah `pd.read_csv()`, kemudian simpan kedalam variable baru bernama `rice_new`. Dengan menggunakan parameter `index_col`, jadikan kolom `receipt_id` sebagai index baris nya!
2. `pandas.DataFrame.head(n)` dapat digunakan untuk menampilkan sebagian data teratas, dengan asumsi bahwa nilai `n` adalah jumlah baris yang ingin kita tampilkan. Silahkan set `head()` dengan nilai `n=8` pada data `rice_new`, kemudian lihat apa yang terjadi! 
3. Lawan dari `head()` adalah `tail()`. Method `tail()` akan menampilkan data dari urutan paling bawah. Silahkan set `tail()` dengan nilai `n=4`, kemudian lihat apa yang terjadi!

In [1556]:
# code here
rice_new = pd.read_csv('../materi_utama/data_input/rice.csv', index_col="receipt_id")
rice_new.head(8)

Unnamed: 0  receipts_item_id    purchase_time category  \
receipt_id                                                           
9622257              1          32369294  7/22/2018 21:19     Rice   
9446359              2          31885876  7/15/2018 16:17     Rice   
9470290              3          31930241  7/15/2018 12:12     Rice   
9643416              4          32418582   7/24/2018 8:27     Rice   
9692093              5          32561236  7/26/2018 11:28     Rice   
9504155              6          32030785  7/17/2018 18:05     Rice   
9822589              7          32935097  7/29/2018 18:18     Rice   
9706953              8          32593606  7/25/2018 12:48     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
9622257            Rice  supermarket 128,000.000         0         1   2018-07  
9446359            Rice   minimarket 102,750.000         0         1   2018-07  
9470290            Rice  supermarket  64,000.000         0         3   2018-07  
9643416            Rice   minimarket  65,000.000         0         1   2018-07  
9692093            Rice  supermarket 124,500.000         0         1   2018-07  
9504155            Rice   minimarket  63,500.000         0         1   2018-07  
9822589            Rice  supermarket  66,500.000         0         1   2018-07  
9706953            Rice   minimarket  62,500.000         0         1   2018-07

In [1557]:
rice_new.tail(4)

Unnamed: 0  receipts_item_id     purchase_time category  \
receipt_id                                                            
5598782          11997          16999147   12/2/2017 14:12     Rice   
5735850          11998          17434503  12/13/2017 19:17     Rice   
5678748          11999          17317935   12/8/2017 22:04     Rice   
5702411          12000          17341559  12/10/2017 17:04     Rice   

           sub_category       format  unit_price  discount  quantity yearmonth  
receipt_id                                                                      
5598782            Rice   minimarket  64,000.000         0         1   2017-12  
5735850            Rice  supermarket  59,990.000      3000         1   2017-12  
5678748            Rice   minimarket  59,500.000      3000         1   2017-12  
5702411            Rice   minimarket  59,500.000      3000         1   2017-12

___

# Tipe Data

Dataframe terdiri dari beberapa **Series** (mengacu pada satu kolom). Dalam satu series harus memiliki satu tipe data yang sama. `pandas`akan mencoba untuk infer tipe data dari masing-masing Series, tapi tidak selalu benar.

## Check Tipe Data

Cara cek tipe data: `dtypes` atau `.info()` untuk lebih lengkapnya

🔎 **Method vs. Atribut**

- Secara fisik/terlihat mata, method selalu diikuti dengan tanda kurung ()
Contoh : head(). tail(), read_csv()
- Secara fisik/terlihat mata, atribut tidak diikuti oleh tanda kurung
Contoh : dtypes
- Didalam sebuah method, nilai parameter itu bisa diganti-ganti
Contoh : head(n=) -> parameter n bisa diganti/disesuaikan dengan jumalh baris yang mau ditampilkan
- Pada sebuah atribut, penggunaaan apa adanya/tidak ada nilai yang bisa diganti ganti

In [1558]:
# read data rice
rice = pd.read_csv('../materi_utama/data_input/rice.csv', index_col=0)

In [1559]:
# check tipe data
rice.dtypes

receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

In [1560]:
# code here (inspect data types with .info())
rice.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 1 to 12000
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   receipt_id        12000 non-null  int64  
 1   receipts_item_id  12000 non-null  int64  
 2   purchase_time     12000 non-null  object 
 3   category          12000 non-null  object 
 4   sub_category      12000 non-null  object 
 5   format            12000 non-null  object 
 6   unit_price        12000 non-null  float64
 7   discount          12000 non-null  int64  
 8   quantity          12000 non-null  int64  
 9   yearmonth         12000 non-null  object 
dtypes: float64(1), int64(4), object(5)
memory usage: 1.0+ MB


✏️ **Notes:**

- `int64`: 
- `float64`: 
- `object`: 

### Knowledge check: `.dtypes` and pandas attributes
Salin kode program di bawah ini, kemudian jalankan. 
```
x = [2019, 4, 'data science']
x.dtypes

```
Apa yang terjadi?

In [1561]:
# code here
x = [2019, 4, 'data science']
# x.dtypes

Ketikkan perintah `type(x)`, maka apa perbedaannya dengan sebelumnya?

In [1562]:
# code here
type(x)

list

> Kesimpulan: dtypes hanya digunakan pada struktur data `DataFrame`, sementara objek `x` tadi tidak bertipe DataFrame melainkan `list`

## ❓ Knowledge Check: Summary Day 2

Pilihlah jawaban yang tepat dengan memberikan tanda centang pada kotak. Jawaban bisa lebih dari satu.

1. Manakah code assignment yang akan menimbulkan **error** ketika dijalankan?

    - [ ] `false = 0` 
    - [X] `0false = 0` 
    - [X] `False = 0` 
    - [ ] `false0 = 0`

2. Pada `pandas`, dikenal istilah `Series` yang mengacu pada sebuah ...

    - [ ] tabel 
    - [ ] baris 
    - [X] kolom 
    - [ ] nilai pada tabel

3. Bagaimana cara mengecek tipe data untuk setiap kolom pada sebuah `DataFrame`?

    - [ ] `pd.read_csv('path/data', index_col= )` 
    - [ ] `df.head()` atau `df.tail()` 
    - [ ] `df.dtypes()` 
    - [X] `df.dtypes`  
    - [ ] `type()` 

**END OF DAY 2**
___
**START OF DAY 3**

## Categorical and Numerical Variables

Tipe data categorical adalah tipe data yang memiliki karakteristik dimana nilainya dapat berulang pada sebuah Series di dataframe. 

Dua alasan mengapa kita perlu menggunakan tipe data categorical:

1. Dari sisi "business perspective", hal ini dapat menginformasikan dan memandu seorang Analyst pada pertanyaan seperti metode statistik atau tipe plot mana yang digunakan untuk mengolah data.

2. Dari sisi teknikal, ketika kita bekerja dengan tipe data categorical pada pandas, hal ini akan jauh menghemat memori dan menambah kecepatan komputasional.

Dikutip dari official documentation:

> Categoricals are a pandas data type corresponding to categorical variables in statistics. A categorical variable takes on a limited, and usually fixed, number of possible values (categories; levels in R). Examples are gender, social class, blood type, country affiliation or rating via Likert scales.

Untuk mengubah tipe data ke categorical pada pandas, Anda dapat melakukannya dengan method `astype()` berikut:

**Formula** 
```
df['column_name'] = df['column_name'].astype('new_data_types')
```

**Contoh**
```
employees['marital_status'] = employees['marital_status'].astype('category')
```

Perhatikan dataframe baru di bawah ini untuk contoh lain dari tipe data pada `pandas`!

In [1563]:
rice[{'category','sub_category'}] 

category sub_category
1         Rice         Rice
2         Rice         Rice
3         Rice         Rice
4         Rice         Rice
5         Rice         Rice
...        ...          ...
11996     Rice         Rice
11997     Rice         Rice
11998     Rice         Rice
11999     Rice         Rice
12000     Rice         Rice

[12000 rows x 2 columns]

In [1564]:
employees = pd.DataFrame({
    'name': ['Anita', 'Brian'],
    'age': [34, 29],
    'joined': [pd.Timestamp('20190410'), pd.Timestamp('20171128')],
    'degree': [True, False],
    'hourlyrate': [35.5, 29],
    'division': ['HR', 'Product']
})
employees.dtypes

name                  object
age                    int64
joined        datetime64[ns]
degree                  bool
hourlyrate           float64
division              object
dtype: object

- `name` [`object`]: store text values
- `age` [`int`]: integer values
- `joined` [`datetime`]: date and time values
- `degree` [`bool`]: True/False values
- `hourlyrate` [`float`]: floating point values
- `division` [`object`]: store text values

❓Perhatikan dataframe `employees`. Adakah kolom yang tipe datanya belum tersimpan dengan benar? Ubahlah ke tipe data yang tepat dengan menggunakan method `astype()`!

Kolom yang perlu diubah tipe datanya:

In [1565]:
# ubah tipe data division
employees['division'] = employees['division'].astype('category')

In [1566]:
# check tipe data
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   name        2 non-null      object        
 1   age         2 non-null      int64         
 2   joined      2 non-null      datetime64[ns]
 3   degree      2 non-null      bool          
 4   hourlyrate  2 non-null      float64       
 5   division    2 non-null      category      
dtypes: bool(1), category(1), datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 320.0+ bytes


___

### Rangkuman Data Types

Fokus pada kolom Pandas dtype dan usage-nya:

| Pandas dtype  | Python type  | NumPy type                                                     | Usage                                        |
|---------------|--------------|----------------------------------------------------------------|----------------------------------------------|
| object        | str or mixed | string_, unicode_, mixed types                                 | Text or mixed numeric and non-numeric values |
| int64         | int          | int_, int8, int16, int32, int64, uint8, uint16, uint32, uint64 | Integer numbers                              |
| float64       | float        | float_, float16, float32, float64                              | Floating point numbers                       |
| bool          | bool         | bool_                                                          | True/False values                            |
| datetime64    | NA           | datetime64[ns]                                                 | Date and time values                         |
| timedelta[ns] | NA           | NA                                                             | Differences between two datetimes            |
| category      | NA           | NA                                                             | Finite list of text values                   |

Referensi: [Overview of Pandas Data Types](https://pbpython.com/pandas_dtypes.html)

___

____

# Exploratory Data Analysis I

Exploratory Data Analysis (**EDA**) mengacu pada proses melakukan investigasi awal pada data, seringkali dengan tujuan untuk lebih mengenal dengan karakteristik data tertentu. EDA dilakukan dengan bantuan ringkasan statistik dan teknik grafis sederhana untuk melihat struktur data yang kita miliki.

EDA:
- Cek data dengan melihat struktur/bentuk data
- Cek data dengan melihat statistika data
- Cek data dengan memvisualisasikan data


Beberapa tools sederhana pada `pandas` yang dapat digunakan untuk melakukan EDA adalah sebagai berikut:
- `.head()` and `.tail()`
- `.describe()`
- `.shape` and `.size`
- `.axes`
- `.dtypes`

## `head()` and `tail()`

- `head(n)` digunakan untuk inspeksi `n` data teratas dari sebuah dataframe.
- `tail(n)` digunakan untuk inspeksi `n` data terbawah dari sebuah dataframe.

## `.describe()`

Method `describe()` menampilkan 8 ringkasan statistika deskriptif. Secara default menampilkan ringkasan untuk kolom numerik. 

Ringkasan statistika yang dimaksud adalah sebagai berikut:
- Count: banyaknya baris pada dataframe
- Mean: rata-rata nilai
- Standard Deviation: jarak rata-rata antara data ke mean (titik pusat data)
- Minimum Value: nilai terkecil dari keseluruhan data
- 25th Percentile (Q1)
- 50th Percentile (Q2/Median)
- 75th Percentile (Q3)
- Maximum Value: nilai terbesar dari keseluruhan data

In [1567]:
# deskripsi dari data rice


✏️ **Insight** : 
- ...
- ...
- ...

Ringkasan seperti count, mean, std, min, max, dan quantile() memiliki method nya masing-masing:

In [1568]:
# code here


___

Kita bisa menambahkan parameter `include` ataupun `exclude` pada `describe()` untuk melihat statistika deskriptif dari variable non-numeric:

In [1569]:
# code here


✏️ **Insight** : 
- ...
- ...
- ...

## `shape` dan `size`

`shape` dan `size` adalah atribut dari sebuah dataframe yang memberikan informasi terkait dimensi data dan ukuran data

In [1570]:
# check shape


✏️ Notes : `shape` akan mengembalikan baris dan kolom (baris, kolom)

In [1571]:
# code here


In [1572]:
# code here


___

In [1573]:
# check size 


___

## `axes`

`axes` adalah atribut dataframe yang memberikan informasi terkait index dataframe (baik index kolom maupun index baris)

In [1574]:
# check axes


✏️ Notes : 

In [1575]:
# code here


In [1576]:
# code here


___

## `dtypes`

Atribut `dtypes` digunakan untuk melakukan inspeksi tipe data.

Karakteristik tipe data `category` :

- Dapat dikelompokkan menjadi beberapa kelompok (category)
- Berulang

Kita bisa menggunakan method berikut untuk mengidentifikasi kolom mana yang cocok untuk disimpan ke tipe data `category`

- `.unique()` to see unique values of a Series
- `.nunique()` to see number of unique values of a Series or DataFrame

Mari kita cek kembali tipe data pada object `rice`. Manakah yang seharusnya memiliki tipe data category?

In [1577]:
rice.dtypes

receipt_id            int64
receipts_item_id      int64
purchase_time        object
category             object
sub_category         object
format               object
unit_price          float64
discount              int64
quantity              int64
yearmonth            object
dtype: object

> **Kolom category:** ...

> **Kolom datetime64:** ...

> **Kolom object:** ...

In [1578]:
# ubah tipe data data data rice


## Knowledge Check: Data types

Misalkan saja kita memiliki sebuah DataFrame dengan nama `inventory`.

1. Ketika menjalankan perintah `inventory.dtypes`, maka pandas akan menampilkan tipe data pada setiap kolom (series). Manakah kolom dibawah ini yang mungkin memiliki tipe data yang salah?
 - [ ] `units_instock`: int64 
 - [ ] `discount_price`: int64
 - [ ] `item_name`: object --> seharusnya diubah menjadi tipe data `category`
 - [ ] `units_sold`: object --> harusnya merupakan integer (`int64`)
 
 
2. Kita ingin mengetahui jumlah kolom yang terdapat pada `inventory` DataFrame, manakah dibawah ini perintah yang tepat untuk menampilkan jumlah kolom `inventory`? Pilih beberapa kode program yang mungkin!
 - [ ] `print(len(inventory.columns))`
 - [ ] `print(inventory.shape[1])`
 - [ ] `print(len(inventory.axes[1]))`

In [1579]:
inventory = pd.DataFrame({
    'units_instock': [50, 40, 30],
    'discount_price': [15, 5, 7],
    'item_name': ['bawang', 'garam', 'gula'],
    'unit_sold': ['123', '456', '789']
})
inventory.dtypes

units_instock      int64
discount_price     int64
item_name         object
unit_sold         object
dtype: object

___

# Indexing and Subsetting with Pandas

Indexing digunakan untuk memilih dan mengambil sebagian data yang hanya diperlukan dalam proses analisa data yang sedang dikerjakan. Contohnya:
- Compare sales pada tahun 2018 vs 2019
- Identifikasi peluang penjualan pada segment pasar (ex : Wholesale vs Retail)
- Melihat quarter terbaik untuk setiap tahun yang dapat digunakan untuk tujuan promosi
- dan sebagainya

Untuk melakukan indexing dan subsetting (slicing data) pada `pandas`, kita dapat menggunakan cara sebagai berikut:
- `head()` and `tail()`  
- `select_dtypes()`  
- Using `.drop()` 
- The `[]` operator
- `.loc`  
- `.iloc`
- Conditional subsetting

## `select_dtypes()`

Method `select_dtypes()` digunakan untuk memilih **kolom** sesuai dengan tipe datanya. Ada 2 parameter yang dapat digunakan di dalam method `select_dtypes()` yaitu parameter `include` dan `exclude` (seperti pada `describe()`.

Misal:
- parameter `include = 'category'` artinya kita memilih semua kolom dengan tipe data 'category'
- sebaliknya, ketika menggunakan parameter `exclude = 'category'` maka kolom-kolom dengan tipe data selain 'category' akan ditampilkan.

In [1580]:
# code here


❓ Bagaimana cara Anda untuk menyeleksi kolom dengan tipe data `int64` dan `float64` untuk ditampilkan?

In [1581]:
# code here


## `drop()`

Method `drop()` digunakan untuk membuang baris atau kolom yang tidak ingin digunakan untuk tujuan analisis. Secara default method ini akan menghapus baris

### Hapus berdasarkan index baris

**Hapus index baris 1 pada DataFrame**

❓ Hapus baris ke 2 dan 4 pada data

In [1582]:
# code here


### Hapus berdasarkan index kolom

**Hapus index kolom `receipt_item_id`**

In [1583]:
# drop columns receipts_item_id


❓ Misal ingin hapus 3 kolom berikut:
- `category` dan `sub_category` karena kita tahu keseluruhan data `rice` merupakan transaksi Rice
- `yearmonth` karena sudah termuat di kolom `purchase_time`

In [1584]:
# code here


___

 ✏️ **NOTE:** Method `drop()` tidak mengubah objek dataframenya. Apabila ingin mengubah objek semulanya:
- Melakukan assignment kembali dengan nama objek yang sama, atau
- Menambahkan parameter inplace

___

## Slicing: **`[]` operator**

Digunakan untuk melakukan subsetting dengan cara mengiris (slicing) index pada dataframe. Formula penulisannya adalah `[start:end]` dengan mengikuti aturan indexing pada python (dimulai dari 0) dimana `start` inclusive dan `end` exclusive.

Dengan menggunakan metode slicing, silahkan tampilkan baris ke 1 dan ke 2

In [1585]:
# slice baris ke 1 dan 2


❓ Dengan menggunakan metode slicing, silahkan tampilkan baris ke 10 sampai ke 15

In [1586]:
# code here


### Knowledge Check: Slicing

Dengan memperhatikan `end` exclusive pada metode indexing, tampilkan baris ke 8 sampai ke 12 pada data household. Pilih jawaban yang tepat dibawah ini!

- [ ] `rice[7:12]`
- [ ] `rice[8:12]`
- [ ] `rice[7:13]`
- [ ] `rice[8:13]`

In [1587]:
# code here


___

## `.iloc` dan `.loc`

Dengan menggunakan `.iloc` dan `loc` kita dapat melakukan pengirisan pada index **baris dan kolom**. 

Perbedaan yang mendasar dari kedua operator ini adalah:
- `.iloc` merujuk pada lokasi **index** baris atau kolomnya sehingga harus **integer**, sedangkan
- `.loc` merujuk pada lokasi **nama** baris atau kolomnya

**Mari berfokus pada .iloc terlebih dahulu**

> Syntax: `df.iloc[baris, kolom]` 

In [1588]:
# menggunakan method .head() 


In [1589]:
# menggunakan slicing


In [1590]:
# menggunakan perintah .iloc


Misal ingin mengambil nilai spesifik, yaitu `2018-07-22 21:19:00` pada baris pertama dan kolom `purchase_time`, maka:

In [1591]:
# code here


❓Mengambil index ke-2 sampai 5 dan kolom dengan `receipt_item_id` sampai `purchase_time`

In [1592]:
# code here


✏️ **Note :** parameter colon (:) digunakan untuk menampilkan semua data baik pada baris ataupun kolom

___

*Mari kita pergi ke .loc**

> Syntax: `df.loc[baris, kolom]` 

In [1593]:
# set receipt_id menjadi label nama baris
rice_new = pd.read_csv("data_input/rice.csv", index_col=1)
rice_new = rice_new.drop('Unnamed: 0', axis=1)
rice_new.head()

FileNotFoundError: [Errno 2] No such file or directory: 'data_input/rice.csv'

Menggunakan `.loc`, kita bisa mengambil baris dan kolom berdasarkan namanya. Contoh kita ingin mengambil baris dengan index baris 1 dan 3, serta mengambil semua kolom.

In [ ]:
# code here


❓ Mengambil baris dengan `receipt_id` 9643416 dan 932907, serta kolom `format` sampai `quantity`

___

## Dive Deeper:
Baca data `companies.csv`dan gunakan `index_col=1`, kemudian tampilkan beberapa baris menggunakan method `head()` atau `tail()`

Dengan menggunakan metode indexing, jawab pertanyaan dibawah ini!

1. Tampilkan lokasi kantor *Li and Partners*?
2. Tampilkan nilai `Returns` yang dihasilkan oleh PT. Algoritma Data Indonesia?
3. Tampilkan nilai `Forecasted Growth` untuk Palembang Konsultansi?

In [ ]:
# code here


___

## ❓ Knowledge Check: Summary Day 3

Pilihlah jawaban yang tepat dengan memberikan tanda centang pada kotak.

1. Statistika deskriptif dari hasil `.describe()` yang digunakan untuk mengetahui **persebaran data** adalah ...

    - [ ] count --> frekuensi
    - [ ] mean --> rata-rata
    - [ ] std
    - [ ] median (50%) --> nilai tengah

2. Code untuk mengetahui **banyaknya kolom** pada dataframe df:

    - [ ] `df.shape[1]`
    - [ ] `df.shape(1)`
    - [ ] `df.axes[1]`
    - [ ] `df.axes(1)`

3. Cara untuk melakukan subsetting berdasarkan **index baris maupun kolom** adalah dengan menggunakan ...

    - [ ] `.select_dtypes()` --> subset kolom berdasarkan type data tertentu
    - [ ] `.drop()` --> membuang baris atau kolom
    - [ ] Slicing operator `[]` --> index baris saja
    - [ ] `.iloc` --> subset berdasarkan index baris maupun index kolom
    - [ ] `.loc` --> subset bedasarkan label baris maupun label kolom
    
4. Pada dataframe df, saya ingin coba **menghapus kolom col1** namun saya ingin perubahannya **tersimpan** pada objek semula. Cara yang tepat adalah ...

    - [ ] `df.drop('col1')` --> kode ini menghapus index
    - [ ] `df.drop(columns='col1')` --> tidak melakukan perubahan
    - [ ] `df = df.drop(columns='col1')` --> cara assignment ke data awal
    - [ ] `df.drop(columns='col1', inplace=True)` --> menggunakan parameter inplace=True
    
5. Misal ada sebuah list sebagai berikut:

    ```{python}
    nama = ["Dwi", "Tomy", "Cut", "Fafil", "Python"]
    ```
    
    Bagaimana cara slicing yang tepat untuk mengambil nilai "Dwi", "Tomy", dan "Owenn"?

    - [ ] `nama[3]`
    - [ ] `nama[:2]`
    - [ ] `nama[:3]`
    - [ ] `nama[:-3]`

___

## Conditional Subsetting

Selain menggunakan `.loc` dan `.iloc`, kita dapat melakukan subsetting berdasarkan kondisi tertentu. Misal pada dataframe `rice`, kita ingin mengambil beberapa data dengan kondisi sebagai berikut:

- Transaksi yang terjadi di supermarket: `.format == 'supermarket'`
- Transaksi dengan produk yang harganya lebih dari sama dengan 200000: `.unit_price >= 200000`
- Transaksi dimana kuantitas tidak sama dengan 0: `.quantity != 0`

Syntax penulisan untuk conditional subsetting adalah:

**`df[df['column_name'] <comparison_operator> <value>]`**

atau

**`df[df.column_name <comparison_operator> <value>]`**

Contoh comparison_operator adalah seperti `==`, `!=`, `>`, `>=`, `<`, `<=`.

Menampilkan data yang memiliki `format` Supermarket!

```
rice[rice['format'] == 'supermarket']

kondisi --> rice['format'] == 'supermarket'
```

In [ ]:
# format supermarket


Menampilkan data yang memiliki `unit_price` lebih besar sama dengan 200.000

In [ ]:
# unit_price >= 200000


❓Tampilkan transaksi yang harganya (`unit_price`) kurang dari 100.000

In [ ]:
# code here


❓ Tampilkan transaksi selain di Supermarket (kolom `format`)!

In [ ]:
# code here


## Multiple Condition

Kita juga dapat menggunakan operator `&` (AND) dan `|` (OR) untuk melakukan subsetting lebih dari 1 kondisi. Misalnya kita ingin melihat data penjualan dari seorang pegawai bernama Moana yang jumlahnya lebih dari 5000, maka kita dapat menggunakan syntax:
```
sales[(sales.salesperson == 'Moana') & (sales.amount > 5000)]
```

Untuk subsetting dengan kondisi lebih dari 1, setiap kondisi diletakkan **di dalam tanda kurung `()`** atau bisa ditulis dengan syntax berikut:

```
df[(kondisi pertama) operator (kondisi kedua) operator (kondisi ketiga) dan seterusnya...]
```

**Poin:**
- Operator AND: harus semua kondisi terpenuhi dalam satu baris agar muncul
- Operator OR: salah satu kondisi saja sudah terpenuhi maka baris tsb muncul
___
- cond1 = kondisi 1
- cond2 = kondisi 2
```
rice[(cond1)&(cond2)]
rice[(cond1)|(cond2)]
```

**Contoh sederhana penggunaan OR dan AND**

In [ ]:
df = pd.DataFrame({
    'angka': [1,2,3],
    'huruf': ['a','b','c']
})
df

**OPERATOR AND**

Menampilkan baris dengan `angka` 1 dan `huruf` a

In [ ]:
#code here
df[(df['angka'] == 1) & (df['huruf'] == 'a')]

Menampilkan baris dengan `angka` 1 dan `huruf` b

In [ ]:
#code here
df[(df['angka'] == 1) & (df['huruf'] == 'b')]

**OPERATOR OR**

Menampilkan baris dengan `angka` 1 atau `huruf` b

In [ ]:
#code here
df[(df['angka'] == 1) | (df['huruf'] == 'b')]

___

❓ Lakukan subsetting untuk mengambil semua transaksi di `supermarket` yang `unit_price`nya di atas 200,000.?

In [ ]:
# code here


❓ Subset baris yang formatnya `supermarket` dan `minimarket`

In [ ]:
# code here


___

## Dive Deeper

Dengan menggunakan metode conditional subsetting, jawab pertanyaan dibawah ini:

1. Apabila Anda masih ingat, pada `.describe()` kita menemui bahwa terdapat transaksi janggal dengan **`discount` yang bernilai negatif**. Ada berapa transaksi yang janggal tersebut?
2. Perusahaan mendefinisikan pembelian grosir sebagai transaksi dengan `quantity` **minimal 12 units (1 Lusin)**. Berapa transaksi pada **hypermarket** yang dikatakan sebagai pembelian grosir?

*Hint: Gunakan `.shape` untuk mengetahui jumlah baris/kolom*

In [ ]:
# your code here: no 1


In [ ]:
# your code here: no 2


**Bonus Challenge**

Berapa banyak transaksi pada dataframe `rice` yang terjadi pada bulan Juli tahun 2018 baik di `minimarket` maupun `supermarket`?

In [ ]:
# code here


___

# Referencing and Copying

Pada python, tanda sama dengan (`=`) bisa digunakan untuk me-**refer** sebuah Dataframe. Sekarang mari kita lihat kembali dataframe `rice`:

In [ ]:
import pandas as pd
rice = pd.read_csv("data_input/rice.csv", index_col=1)
rice = rice.drop('Unnamed: 0', axis=1)
rice.head()

## Reference

Membuat dataframe baru dengan nama `rice_second` yang menyimpan semua transaksi seperti `rice`. Kemudian misalnya kita ingin update kolom `discount` agar menjadi 15 pada semua barisnya.

In [ ]:
#code here


In [ ]:
# check data rice


In [ ]:
# check data rice_second


In [ ]:
# check memory id rice dan rice_second


## Copying

Membuat dataframe baru dengan nama `rice_second` yang menyimpan semua transaksi seperti `rice`. Kemudian misalnya kita ingin update kolom `discount` agar menjadi 15 pada semua barisnya.

In [ ]:
#code here
rice = pd.read_csv("data_input/rice.csv", index_col=1)
rice.head()

In [ ]:
#code here


In [ ]:
# check data rice


In [ ]:
# check data rice_second


In [ ]:
# check memory id rice dan rice_second


___

# Additional Material

Cara membuat kolom baru
    ```
    df['new_column'] = new_value
    ```

`new_column` merupakan nama kolom yang sebelumnya tidak ada di dataframe `df`.

Kasus: Ingin menghitung `total_price` yaitu hasil perkalian `unit_price` dengan `quantity`

In [ ]:
rice['total_price'] = rice['unit_price'] * rice['quantity']

In [ ]:
rice